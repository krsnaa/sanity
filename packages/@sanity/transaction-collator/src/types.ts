export type Mutation = {
  create: any
  delete: any
  createIfNotExists: any
  createOrReplace: any
  patch: any
}

export type HistoryEvent = {
  type: 'created' | 'edited' | 'published' | 'unpublished' | 'unknown'
  userIds: string[]
  startTime: Date
  endTime: Date,
  documentIDs: string[]
  rev: string
}

export type Transaction = {
  namespace: string
  id: string
  author: string
  mutations: Mutation[]
  documentIDs: string[]
  timestamp: string
}
