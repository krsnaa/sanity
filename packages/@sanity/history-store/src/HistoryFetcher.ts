import { client } from './utils/Client'

export function getHistory(documentIds: string | string[], options: { [key: string]: any } = {}): Promise<any> {
  const ids = Array.isArray(documentIds) ? documentIds : [documentIds]
  const { time, revision } = options

  if (time && revision) {
    throw new Error(`getHistory can't handle both time and revision parameters`)
  }

  const dataset = client.clientConfig.dataset
  let url = `/data/history/${dataset}/documents/${ids.join(',')}`

  if (revision) {
    url = `${url}?revision=${revision}`
  } else {
    const timestamp = time || new Date().toISOString()
    url = `${url}?time=${timestamp}`
  }

  return client.request({ url })
}

export function getTransactions(documentIds: string | string[]): Promise<any> {
  const ids = Array.isArray(documentIds) ? documentIds : [documentIds]
  const dataset = client.clientConfig.dataset
  const url = `/data/history/${dataset}/transactions/${ids.join(',')}`

  return client.request({ url })
}

