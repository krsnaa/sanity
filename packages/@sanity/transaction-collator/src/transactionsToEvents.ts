import {uniq} from 'lodash'
import {HistoryEvent, Transaction, Mutation} from './types'
import {ndjsonToArray} from './utils/ndjsonToArray'

const EDIT_EVENT_TIME_TRESHHOLD_MS = 5 * 1000 * 60 * 5 // 5 minutes

export function transactionsToEvents(
  documentId: string,
  transactions: string | Buffer
): HistoryEvent[] {
  const rawItems = ndjsonToArray(transactions)
  return (
    rawItems
      // Make sure we only deal with ids that are for our document (including the draft)
      .filter(
        (transaction: Transaction) =>
          transaction.documentIDs &&
          (transaction.documentIDs.includes(documentId) ||
            transaction.documentIDs.includes(`drafts.${documentId}`))
      )
      // Turn a transaction into a classified HistoryEvent
      .map(mapToEvents)
      // Chunk and group edit events
      .reduce(reduceEdits, [])
  )
}

function mapToEvents(transaction: Transaction): HistoryEvent {
  const type = mutationsToEventType(transaction.mutations)
  const timestamp = new Date(transaction.timestamp)
  return {
    type,
    rev: transaction.id,
    userIds: [transaction.author],
    startTime: timestamp,
    endTime: timestamp
  }
}

function reduceEdits(
  acc: HistoryEvent[],
  current: HistoryEvent,
  index: number,
  arr: HistoryEvent[]
) {
  const nextEvent = arr[index + 1]
  const skipEvent =
    current.type === 'edited' &&
    nextEvent &&
    nextEvent.type === 'edited' &&
    nextEvent.endTime.getTime() - current.endTime.getTime() < EDIT_EVENT_TIME_TRESHHOLD_MS
  if (skipEvent) {
    // Lift authors over to next event
    nextEvent.userIds = uniq(nextEvent.userIds.concat(current.userIds))
    // Set startTime on next event to be this one if not done already
    // (then startTime and endTime would be different)
    if (current.startTime.getTime() === current.endTime.getTime()) {
      nextEvent.startTime = current.startTime
    }
  } else {
    acc.push(current)
  }
  return acc
}

function mutationsToEventType(mutations: Mutation[]) {
  // Created
  if (
    mutations.length === 2 &&
    mutations[0].createIfNotExists &&
    mutations[0].createIfNotExists._id.startsWith('drafts.') &&
    mutations[1].patch.id.startsWith('drafts.')
  ) {
    return 'created'
  }

  // Published
  if (
    mutations.length === 2 &&
    mutations[0].create &&
    mutations[1].delete &&
    mutations[1].delete.id.startsWith('drafts.')
  ) {
    return 'published'
  }
  // Unpublished
  if (
    mutations.length === 2 &&
    mutations[1].createIfNotExists &&
    mutations[1].createIfNotExists._id.startsWith('drafts.') &&
    mutations[0].delete &&
    !mutations[0].delete.id.startsWith('drafts.')
  ) {
    return 'unpublished'
  }

  // Edited
  if (mutations.some(mut => mut.patch)) {
    return 'edited'
  }

  return 'unknown'
}
