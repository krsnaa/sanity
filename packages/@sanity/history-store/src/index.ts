import { getHistory, getTransactions } from './HistoryFetcher'

const HistoryStore = {
  getHistory,
  getTransactions
}

export { HistoryStore }
