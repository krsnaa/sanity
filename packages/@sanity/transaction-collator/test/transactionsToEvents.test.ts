import {readFileSync} from 'fs'
import {transactionsToEvents} from '../src/index'
const ndJSON = readFileSync('./test/fixtures/transactions.ndjson', 'utf8')
const documentId = '33d64a8a-daaf-41e7-849f-657460c33559'

test('parses ndjson', () => {
  const result = transactionsToEvents(documentId, ndJSON)
  console.log(result)
  expect(result).toMatchSnapshot()
})
