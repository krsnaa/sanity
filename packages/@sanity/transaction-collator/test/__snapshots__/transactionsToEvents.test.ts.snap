// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses ndjson 1`] = `
Array [
  Object {
    "authors": Array [
      "ppJZEsNw6",
    ],
    "endTime": 2019-05-29T08:03:40.026Z,
    "startTime": 2019-05-29T08:03:40.026Z,
    "type": "created",
  },
  Object {
    "authors": Array [
      "ppJZEsNw6",
    ],
    "endTime": 2019-05-29T08:03:55.787Z,
    "startTime": 2019-05-29T08:03:55.787Z,
    "type": "edited",
  },
  Object {
    "authors": Array [
      "ppJZEsNw6",
    ],
    "endTime": 2019-05-29T08:03:57.598Z,
    "startTime": 2019-05-29T08:03:57.598Z,
    "type": "published",
  },
  Object {
    "authors": Array [
      "ppJZEsNw6",
    ],
    "endTime": 2019-05-29T08:06:26.839Z,
    "startTime": 2019-05-29T08:06:26.839Z,
    "type": "unpublished",
  },
  Object {
    "authors": Array [
      "ppJZEsNw6",
    ],
    "endTime": 2019-05-29T08:06:36.860Z,
    "startTime": 2019-05-29T08:06:36.860Z,
    "type": "edited",
  },
  Object {
    "authors": Array [
      "ppJZEsNw6",
    ],
    "endTime": 2019-05-29T08:06:40.080Z,
    "startTime": 2019-05-29T08:06:40.080Z,
    "type": "published",
  },
  Object {
    "authors": Array [
      "ppJZEsNw6",
    ],
    "endTime": 2019-05-29T08:06:46.378Z,
    "startTime": 2019-05-29T08:06:45.347Z,
    "type": "edited",
  },
  Object {
    "authors": Array [
      "ppJZEsNw6",
    ],
    "endTime": 2019-05-29T09:11:41.204Z,
    "startTime": 2019-05-29T09:11:40.172Z,
    "type": "edited",
  },
]
`;
